# Quantum Coin Toss Using a Hadamard Gate

---

## Introduction

### 1. Problem Statement
Simulate a quantum coin toss by initializing a single qubit in the computational basis state $|0\rangle$, applying a Hadamard gate to create an equal superposition, measuring the qubit in the computational basis, and executing the circuit for 1000 shots. Interpret measurement outcome 0 as Heads and outcome 1 as Tails, then calculate the empirical probability of each outcome.

---

### 2. Concept Explanation

#### Initial State
The single-qubit quantum system begins initialized in the ground computational basis state:
$$|0\rangle = \begin{pmatrix} 1 \\ 0 \end{pmatrix}$$

#### Equal Superposition via the Hadamard Gate
The single-qubit Hadamard gate ($H$) transforms deterministic computational basis states into symmetric equal-superposition states:
$$H = \frac{1}{\sqrt{2}} \begin{pmatrix} 1 & 1 \\ 1 & -1 \end{pmatrix}$$
Applying the Hadamard gate to the initial ground state $|0\rangle$ produces the $|+\rangle$ state:
$$H|0\rangle = \frac{|0\rangle + |1\rangle}{\sqrt{2}} = \begin{pmatrix} \frac{1}{\sqrt{2}} \\ \frac{1}{\sqrt{2}} \end{pmatrix}$$

#### Measurement Collapse and Quantum Randomness
Unlike classical deterministic physics, quantum mechanics introduces fundamental randomness upon measurement. When a qubit in superposition $\frac{|0\rangle + |1\rangle}{\sqrt{2}}$ is measured in the standard computational basis $\{|0\rangle, |1\rangle\}$, its quantum state irreversibly collapses to either $|0\rangle$ or $|1\rangle$.

#### Theoretical Outcome Probabilities (Born Rule)
According to the **Born rule**, the probability of observing basis state $|x\rangle$ is the squared magnitude of its probability amplitude:
$$P(0) = |\langle 0 | \psi \rangle|^2 = \left|\frac{1}{\sqrt{2}}\right|^2 = \frac{1}{2} = 0.5 \quad (50\%)$$
$$P(1) = |\langle 1 | \psi \rangle|^2 = \left|\frac{1}{\sqrt{2}}\right|^2 = \frac{1}{2} = 0.5 \quad (50\%)$$

#### Finite Sampling and Statistical Fluctuations
When executing a quantum circuit for a finite number of shots ($N = 1000$), the observed frequencies will fluctuate around the theoretical mean of $50\%$. The standard deviation of the sampling error for a binomial distribution is:
$$\sigma = \sqrt{\frac{p(1-p)}{N}} = \sqrt{\frac{0.5 \times 0.5}{1000}} \approx 0.0158 \quad (1.58\%)$$
Consequently, slight variations around $0.5$ in 1000 shots are completely natural and reflect genuine statistical sampling noise.

#### Coin Toss Outcome Mapping
The measurement outcomes are mapped to classical coin-toss events:
- **Outcome `0`** $\longrightarrow$ **Heads**
- **Outcome `1`** $\longrightarrow$ **Tails**

---

### 3. Experiment Setup
- **Number of Qubits**: 1
- **Quantum Gate**: Hadamard ($H$)
- **Measurement**: Full computational basis measurement (`measure_all()`)
- **Number of Shots**: 1000
- **Simulator**: Qiskit Aer `AerSimulator()`
- **Possible Outcomes**: 0 (Heads) and 1 (Tails)

---

### 4. Tools and Libraries
- **Qiskit**: Open-source quantum SDK used to construct and visualize the quantum circuit (`QuantumCircuit`).
- **Qiskit Aer (`AerSimulator`)**: High-performance simulator used to perform shot-based quantum measurements and sample experimental outcomes.
- **Python / NumPy**: Used for aggregating measurement counts and computing empirical probabilities.

---

## Implementation Setup

### Dependencies and Imports
Install and import the necessary modules from Qiskit and NumPy.

In [2]:
!pip install qiskit qiskit-aer -q

In [3]:
import random
import numpy as np

from qiskit import QuantumCircuit
from qiskit.quantum_info import Statevector
from qiskit_aer import AerSimulator

/var/folders/yp/078pyxw11mq6g7sd_2ty18z80000gn/T/ipykernel_27322/2087795024.py:4: DeprecationWarning: Using Qiskit with Python 3.9 is deprecated as of the 2.1.0 release. Support for running Qiskit with Python 3.9 will be removed in the 2.3.0 release, which coincides with when Python 3.9 goes end of life.
  from qiskit import QuantumCircuit


### Helper Function
Define a reusable helper function `show_state(circuit, title)` to inspect statevectors and circuit drawings.

In [4]:
def show_state(circuit, title):
    state = Statevector.from_instruction(circuit)
    print("="*60)
    print(title)
    print(circuit.draw())
    print("\nStatevector:")
    print(state)
    print("="*60)
    return state

---

## Create the Quantum Coin Circuit

### Apply the Hadamard Gate
Initialize a single-qubit quantum circuit (`QuantumCircuit(1)`) in the state $|0\rangle$ and apply a Hadamard gate using `coin.h(0)` to prepare an equal superposition of $|0\rangle$ and $|1\rangle$.

### Measure the Qubit
Attach measurement operations using `coin.measure_all()` to measure the qubit in the computational basis and store the result in a classical bit.

In [5]:
coin = QuantumCircuit(1)

coin.h(0)

coin.measure_all()

print(coin.draw())

        ┌───┐ ░ ┌─┐
     q: ┤ H ├─░─┤M├
        └───┘ ░ └╥┘
meas: 1/═════════╩═
                 0 


---

## Run the Simulation

Execute the quantum coin-toss circuit for 1000 shots using Qiskit Aer's `AerSimulator()` to sample measurement outcomes.

In [6]:
sim = AerSimulator()

job = sim.run(coin, shots=1000)

result = job.result()

counts = result.get_counts()

print("Measurement Counts:\n")
print(counts)

Measurement Counts:

{'1': 493, '0': 507}


---

## Calculate Heads and Tails Probabilities

Extract the counts for outcome `0` (Heads) and outcome `1` (Tails), normalize by the total number of shots ($N = 1000$), and compute the empirical probabilities.

In [7]:
shots = sum(counts.values())

prob_0 = counts.get('0',0)/shots

prob_1 = counts.get('1',0)/shots

print("Probability of Heads (0):",prob_0)

print("Probability of Tails (1):",prob_1)

Probability of Heads (0): 0.507
Probability of Tails (1): 0.493


---

## Interpret the Results

### Comparison of Observed vs. Theoretical Outcomes

| Outcome | Bit Value | Observed Counts ($N = 1000$) | Empirical Probability | Theoretical Probability | Deviation |
| :---: | :---: | :---: | :---: | :---: | :---: |
| **Heads** | `0` | 507 | 0.507 ($50.7\%$) | 0.500 ($50.0\%$) | $+0.007$ ($+0.7\%$) |
| **Tails** | `1` | 493 | 0.493 ($49.3\%$) | 0.500 ($50.0\%$) | $-0.007$ ($-0.7\%$) |

### Statistical Analysis
- The observed distribution yields 507 Heads ($50.7\%$) and 493 Tails ($49.3\%$).
- The deviation of $\pm 0.7\%$ from the exact theoretical probability of $50\%$ falls comfortably within the expected $1\sigma$ standard sampling error ($\sigma \approx 1.58\%$).
- This confirms that the Hadamard gate acts as a balanced quantum random number generator producing true quantum coin-toss behavior.

---

## Conclusion

- The **Hadamard gate** transforms the ground state $|0\rangle$ into an equal superposition $\frac{|0\rangle + |1\rangle}{\sqrt{2}}$, producing an approximately balanced quantum coin toss.
- Upon computational basis measurement, the state collapses probabilistically to either $|0\rangle$ (Heads) or $|1\rangle$ (Tails), with equal theoretical probability $P(0) = P(1) = 0.5$.
- In the 1000-shot simulation, the observed empirical probabilities ($0.507$ for Heads and $0.493$ for Tails) closely match the theoretical expectation.
- Finite sampling causes minor statistical fluctuations around $0.5$, which are expected to diminish as the number of shots increases according to the law of large numbers.